File: 202-happy-number/happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        def mutate(num):
            res = 0
            while num > 0:
                res += (num % 10) ** 2
                num //= 10
            return res
        slow = fast = n
        while True:
            slow = mutate(slow)
            fast = mutate(mutate(fast))
            if slow == fast:
                slow = n
                while slow != fast:
                    slow = mutate(slow)
                    fast = mutate(fast)
                return slow == 1

File: 202-happy-number/happy-number.cpp
class Solution {
private:
    int calcSum(int n) {
        int res = 0;
        while (n) {
            res += (n % 10) * (n % 10);
            n /= 10;
        }
        return res;
    }
public:
    bool isHappy(int n) {
        int slow = n;
        int fast = n;
        while (true) {
            slow = calcSum(slow);
            fast = calcSum(calcSum(fast));
            if (slow == fast) {
                return slow == 1;
            }
        }
    }
};